src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -117,18 +118,29 @@ private T runWithRetries(byte[] key, int redirections, boolean tryRandomNode, bo
       }
 
       return execute(connection);
-    } catch (JedisConnectionException jce) {
-      if (tryRandomNode) {
-        // maybe all connection is down
-        throw jce;
-      }
 
+    } catch (JedisNoReachableClusterNodeException jnrcne) {
+      //Not sure if JedisNoReachableClusterNodeException exception should extends JedisConnectionException, it's not connection exception
+      //Remove this if hierarchy reworked and no need to make exception from JedisConnectionException clause
+      throw jnrcne;
+    } catch (JedisConnectionException jce) {
       // release current connection before recursion
       releaseConnection(connection);
       connection = null;
 
-      // retry with random connection
-      return runWithRetries(key, redirections - 1, true, asking);
+      if (redirections <= 1) {
+        //We need this because if node is not reachable anymore - we need to finally initiate slots renewing,
+        //or we can stuck with cluster state without one node in opposite case.
+        //But now if maxRedirections = 1 or 2 we will do it too often. For each time-outed request.
```

Comment:
> Do you really want to get rid of cluster rediscovering on timeouts if you have maxRedirects=1?

I'm kind of torn. I don't know what the right thing is here. :(

I like the change to avoid using `tryRandomNode` -- for my use case (and for yours too, it sounds like), it only adds unnecessary hops within the cluster that aren't likely to result in success.

The renewSlotCache change adds unnecessary work for the Jedis clients for my use case, because we don't have any slaves to fail over, BUT it's totally possible my use case is weird, and your proposed solution here makes things better for the majority of users, in which case you should ignore me, and I'll try to figure out a way to make this code pluggable (maybe a RetryHandler, or something?) that could allow more flexibility in how clients fail. In any event, I'm just an enthusiast, and have no say in whether or not this gets merged anyway. :)